isobus/include/isobus/isobus/can_parameter_group_builder.hpp

```diff
@@ -0,0 +1,503 @@
+		bool write<char *>(char * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<std::uint8_t const *>(std::uint8_t const * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<std::uint8_t *>(std::uint8_t * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		bool write(char const * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool write(char * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool write(std::uint8_t const * data, bool includeNull)
+		{
+			// Base case.  Write each byte separately so they don't get put in little-endian, which
+			// makes no sense for strings.
+			std::size_t revert = writeOffset;
+			while (*data)
+			{
+				if (!write_bits(data, 8))
+				{
+					writeOffset = revert;
+					return false;
+				}
+				++data;
+			}
+			if (includeNull)
+			{
+				std::uint8_t naught = 0;
+				if (!write_bits(&naught, 8))
+				{
+					writeOffset = revert;
+					return false;
+				}
+			}
+			return true;
+		}
+
+		bool write(std::uint8_t * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool pad(std::size_t bits, bool value = true)
+		{
+			std::uint8_t data = value ? 255 : 0;
+			std::size_t byte = get_write_byte_offset();
+			std::size_t offset = get_write_bit_offset();
+			std::size_t remaining = 8 - offset;
+			std::uint8_t mask = 255 >> remaining;
+			writeOffset += bits;
+			buffer.resize((writeOffset + 7) / 8, 255);
+			for (;;)
+			{
+				buffer[byte] = (buffer[byte] & mask) | (data << offset);
+				if (remaining > bits)
+				{
+					// Weirdly we need a second masking here, to keep untouched bits as `1`.
+					buffer[byte] |= 255 << (bits + offset);
+					break;
+				}
+				else if (remaining == bits)
+				{
+					// Done.
+					break;
+				}
+				else
+				{
+					bits -= remaining;
+					offset = 0;
+					remaining = 8;
+					mask = 0;
+					++byte;
+				}
+			}
+			return true;
+		}
+
+		template <typename T>
+		bool read(T & data)
+		{
+			return read_bits((std::uint8_t *)&data, sizeof (T) * 8);
+		}
+
+		template <typename T>
+		bool read(T & data, std::size_t bits)
+		{
+			// Clear the memory, since we may not be reading the full width.
+			memset(&data, 0, sizeof(T));
+			return read_bits((std::uint8_t *)&data, bits);
+		}
+
+		template <>
+		bool read<bool>(bool & data)
+		{
+			std::uint8_t bits = 0;
+			if (read_bits(&bits, 1))
+			{
+				data = !!bits;
+				return true;
+			}
+			data = false;
+			return false;
+		}
+
+		template <>
+		bool read<char *>(char * & data)
+		{
+			// Read until NULL.
+			return read<std::uint8_t*>((std::uint8_t * &)data);
```

Comment:
It seems like a dangerous assumption that a `uint8_t *` will have null termination here and other places. That's a non-trivial assumption that cannot be enforced. `std::vector<std::uint8_t>` is not a `std::string()::c_str()`

Reply:
Yes it is, and I know how that bad assumption sneaked in (I did a find/replace on `unsigned char` to `uint8_t`).  TBH I was already unhappy with the way strings were handled in here, so I'll have to revisit that entire part.